clang/test/CodeGenSYCL/sycl-device-only-virtual-functions.cpp

```diff
@@ -0,0 +1,145 @@
+// Purpose of this test is to ensure that in SYCL device code mode, we are only
+// emitting those virtual functions into vtable, which are marked with
+// [[intel::device_indirectly_callable]] attribute.
```

Comment:
Implementation also allows emission of functions which are marked with `sycl_device` attribute only, if I understand correctly. That attribute will likely be the only thing we check to decide if we need to emit a virtual function into device code going forward. Therefore, I think that it is important to have it checked as well, i.e. in this test we should replace a couple of `[[intel::device_indirectly_callable]]` occurrences with `SYCL_EXTERNAL` and that should not change the resulting LLVM IR

Reply:
>>That attribute will likely be the only thing we check to decide if we need to emit a virtual function into device code going forward. 
Actually do you want to remove support for  [[intel::device_indirectly_callable]] then ? I can do it here and then make changes downstream as well.

Reply:
> Actually do you want to remove support for `[[intel::device_indirectly_callable]]` then ? I can do it here and then make changes downstream as well.

We may have some internal customers relying on behavior of this attribute, so I would keep it for now. I will work on its removal alongside with the extension implementation